ModelFarm/src/main/java/io/deephaven/modelfarm/ModelFarmBase.java

```diff
@@ -340,8 +340,8 @@ public final boolean awaitTermination(final long timeout, final TimeUnit unit) {
                 timeout == Long.MAX_VALUE ? Long.MAX_VALUE : System.currentTimeMillis() + unit.toMillis(timeout);
         boolean allThreadsTerminated = false;
 
-        while (!allThreadsTerminated && System.currentTimeMillis() < timeoutMillis) {
-            synchronized (ModelFarmBase.this) {
+        synchronized (ModelFarmBase.this) {
+            while (!allThreadsTerminated && System.currentTimeMillis() < timeoutMillis) {
```

Comment:
You should record the current time (or the remaining time) from this line and use that in the `wait` below on L348. Currently you run the risk of an `IllegalArgumentException` if you ask to wait for a negative time, or blocking indefinitely if you ask to wait for zero milliseconds.

Reply:
Good catch.